File: apps/android/M0ProtocolMockClient/src/main/kotlin/com/commandrelay/m0protocol/replay/M0ReplayCursorStore.kt
package com.commandrelay.m0protocol.replay

/**
 * Store abstraction that tracks the latest acknowledged sequence per stream.
 */
public interface M0ReplayCursorStore {
    /**
     * Returns the latest acknowledged sequence for a stream.
     *
     * @param streamId Stream identifier.
     * @return Stored cursor for the stream, or null when unknown.
     */
    public fun lastSeq(streamId: String): Long?

    /**
     * Records an acknowledged sequence for a stream.
     *
     * Implementations should keep the maximum observed value for each stream.
     *
     * @param streamId Stream identifier.
     * @param seq Sequence to acknowledge.
     */
    public fun record(streamId: String, seq: Long)

    /**
     * Returns a complete cursor snapshot for diagnostics and tests.
     *
     * @return Immutable stream-to-sequence cursor map.
     */
    public fun snapshot(): Map<String, Long>
}

/**
 * In-memory [M0ReplayCursorStore] implementation for tests and local protocol mocks.
 *
 * @param seed Initial stream-to-sequence cursor values.
 */
public class InMemoryM0ReplayCursorStore(seed: Map<String, Long> = emptyMap()) : M0ReplayCursorStore {
    private val cursors: MutableMap<String, Long> = seed.toMutableMap()

    /**
     * Returns the latest acknowledged sequence for [streamId].
     */
    override fun lastSeq(streamId: String): Long? = cursors[streamId]

    /**
     * Records [seq] for [streamId], preserving the maximum value seen.
     */
    override fun record(streamId: String, seq: Long) {
        val current = cursors[streamId] ?: 0L
        cursors[streamId] = maxOf(current, seq)
    }

    /**
     * Returns a detached snapshot so callers cannot mutate internal state.
     */
    override fun snapshot(): Map<String, Long> = cursors.toMap()
}

File: apps/ios/M0ProtocolMockClient/Sources/M0ProtocolMockClient/Replay/M0ReplayCursorStore.swift
import Foundation

/// In-memory cursor store that tracks the latest acknowledged sequence per stream.
public actor M0ReplayCursorStore {
    private var cursors: [String: UInt64]

    /// Creates a cursor store with optional seed values.
    /// - Parameter seed: Initial stream-to-sequence cursor values.
    public init(seed: [String: UInt64] = [:]) {
        self.cursors = seed
    }

    /// Returns the latest acknowledged sequence for a stream.
    /// - Parameter streamID: Stream identifier.
    /// - Returns: The stored cursor if one exists.
    public func lastSeq(for streamID: String) -> UInt64? {
        cursors[streamID]
    }

    /// Records an acknowledged sequence, keeping the maximum cursor seen.
    /// - Parameters:
    ///   - streamID: Stream identifier.
    ///   - seq: Sequence to acknowledge.
    public func record(streamID: String, seq: UInt64) {
        let current = cursors[streamID] ?? 0
        cursors[streamID] = max(current, seq)
    }

    /// Returns a complete cursor snapshot for diagnostics and tests.
    /// - Returns: Stream cursor map.
    public func snapshot() -> [String: UInt64] {
        cursors
    }
}
